src/AppChooser/Dialog.vala

```diff
@@ -127,32 +129,51 @@ public class AppChooser.Dialog : Hdy.Window {
         };
         open_button.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
 
+        var top_grid = new Gtk.Grid () {
+            orientation = Gtk.Orientation.VERTICAL,
+            column_spacing = 12,
+            row_spacing = 6,
+            margin_start = 12,
+            margin_end = 12,
+            margin_top = 12
+        };
+
+        top_grid.attach (overlay, 0, 0, 1, 2);
+        top_grid.attach (primary_label, 1, 0);
+        top_grid.attach (secondary_label, 1, 1);
+
+        var window_handle_top = new Hdy.WindowHandle ();
+        window_handle_top.add (top_grid);
+
+        var content_grid = new Gtk.Grid () {
+            orientation = Gtk.Orientation.VERTICAL,
+            column_spacing = 12,
+            row_spacing = 6,
+            margin_start = 12,
+            margin_end = 12
+        };
+
+        content_grid.attach (frame, 0, 0);
+
         var button_box = new Gtk.ButtonBox (Gtk.Orientation.HORIZONTAL) {
             layout_style = Gtk.ButtonBoxStyle.END,
-            margin_top = 12,
+            margin = 12,
             spacing = 6
         };
 
         button_box.add (cancel);
         button_box.add (open_button);
 
-        var grid = new Gtk.Grid () {
-            orientation = Gtk.Orientation.VERTICAL,
-            column_spacing = 12,
-            row_spacing = 6,
-            margin = 12
-        };
+        var window_handle_bottom = new Hdy.WindowHandle ();
+        window_handle_bottom.add (button_box);
 
-        grid.attach (overlay, 0, 0, 1, 2);
-        grid.attach (primary_label, 1, 0);
-        grid.attach (secondary_label, 1, 1);
-        grid.attach (frame, 0, 3, 2);
-        grid.attach (button_box, 1, 4);
+        var box = new Gtk.Box (Gtk.Orientation.VERTICAL, 6);
 
-        var window_handle = new Hdy.WindowHandle ();
-        window_handle.add (grid);
+        box.pack_start (window_handle_top, false, false);
+        box.pack_start (content_grid);
+        box.pack_start (window_handle_bottom, false, false);
```

Comment:
this should be `add()` anyway, `pack_start()` was removed/renamed in Gtk4, also [child properties too](https://docs.gtk.org/gtk4/migrating-3to4.html#stop-using-gtkbox-padding-fill-and-expand-child-properties), so it's better to keep using add() in this case.